Remotion/Web/Development.WebTesting/Utilities/ProcessUtils.cs

```diff
@@ -134,16 +136,103 @@ public static void GracefulProcessShutdown ([NotNull] Process process, TimeSpan
         // Tested with process running with admin rights.
         // If there is an unexpected exception which causes any of the other methods to fail, we have to decide how we want to handle that.
       }
-      
+
       // Try to gracefully close the process
-      if (process.CloseMainWindow())
+      if (CloseMainWindow (process))
       {
         // No exception is thrown when calling .WaitForExit(..) on an already exited process.
         if (process.WaitForExit (timeoutInMilliseconds))
           return;
       }
 
       // Force closing the process
+      KillProcess (process);
+
+      // No exception is thrown when calling .WaitForExit(..) on an already exited process.
+      if (process.WaitForExit (timeoutInMilliseconds))
+        return;
+
+      throw CreateProcessDidNotExitInTimeException (process);
+    }
+
+    /// <summary>
+    /// Tries to shutdown a list of processes gracefully and if they are not closed after <paramref name="timeout"/> it ensures that they are closed.
+    /// </summary>
+    /// <param name="processes">The processes that should be shut down.</param>
+    /// <param name="timeout">The wait timeout after a shutdown has been requested.</param>
+    /// <remarks>
+    /// <p>
+    /// This method will block until the processes have exited. Shutdown procedure is as follows:
+    /// </p>
+    /// <list type="number">
+    /// <item>
+    /// <description>Check if the processes have already exited.</description>
+    /// </item>
+    /// <item>
+    /// <description>Close the main windows of the processes and wait for the processes to exit or the <paramref name="timeout"/> to run out.</description>
+    /// </item>
+    /// <item>
+    /// <description>If the processes are still not closed, kill them and wait for the processes to exit or the <paramref name="timeout"/> to run out.</description>
+    /// </item>
+    /// </list>
+    /// </remarks>
+    public static void GracefulProcessShutdown (IReadOnlyList<Process> processes, TimeSpan timeout)
```

Comment:
Isn't the single-process variant of this method just an overload? Can be do this:
`public static void GracefulProcessShutdown ([NotNull] Process process, TimeSpan timeout)
{
  GracefulProcessShutdown (new[]{process}, timeout);
}`

And, is the single-process overload actually needed in the first place?

Reply:
The single process overload is no longer needed, but I wasn't sure if it needs to be kept. If it is not needed I would remove it.

Reply:
Keep in mind that there could be some dependencies on this, as this is a public class.

Reply:
Well, we do have the DependDB and it would be easy to adapt a broken invocation. But I don't mind keeping the overload.